### PR TITLE
[Multi NPU] Time Improvements to the config reload/load_minigraph commands 

### PR DESCRIPTION
**Pull request by judyjoseph**
**- What I did**
config reload and config load_minigraph commands were taking ~ 5m in in the multi ASIC platforms.**Reduced the total time taken to ~ 1min by parallelizing with python threads.**

**- How I did it**
Started a python thread per NPU to stop/restart the services. 

Currently the threads are started at the same time for all the ASIC's irrespective of whether they are front-end OR back-end ASIC's. I found it ok as we finish the entire activity of config reload in 1 min and when we do config reload usually the device should not be in production and not carrying active traffic.

**- How to verify it**
Verified on a Single ASIC and multi AISC devices. Attaching sample output.

**MULTI ASIC**

**~$ time sudo config reload  -y**
Executing stop of service swss@1...
Executing stop of service swss@0...
Executing stop of service swss@2...
Executing stop of service swss@3...
Executing stop of service swss@4...
Executing stop of service swss@5...
Executing stop of service lldp...
Executing stop of service lldp@1...
Executing stop of service lldp@0...
Executing stop of service lldp@2...
Executing stop of service lldp@3...
Executing stop of service lldp@4...
Executing stop of service lldp@5...
Executing stop of service pmon...
Executing stop of service bgp@0...
Executing stop of service bgp@1...
Executing stop of service bgp@2...
Executing stop of service bgp@3...
Executing stop of service bgp@4...
Executing stop of service bgp@5...
Executing stop of service hostcfgd...
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db.json --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db0.json -n asic0 --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate -n asic0
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db1.json -n asic1 --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate -n asic1
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db2.json -n asic2 --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate -n asic2
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db3.json -n asic3 --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate -n asic3
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db4.json -n asic4 --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate -n asic4
Running command: /usr/local/bin/sonic-cfggen -j /etc/sonic/init_cfg.json -j /etc/sonic/config_db5.json -n asic5 --write-to-db
Running command: /usr/bin/db_migrator.py -o migrate -n asic5
Executing reset-failed of service bgp@0...
Executing reset-failed of service bgp@1...
Executing reset-failed of service bgp@2...
Executing reset-failed of service bgp@3...
Executing reset-failed of service bgp@4...
Executing reset-failed of service bgp@5...
Executing reset-failed of service dhcp_relay...
Executing reset-failed of service hostcfgd...
Executing reset-failed of service hostname-config...
Executing reset-failed of service interfaces-config...
Executing reset-failed of service lldp...
Executing reset-failed of service lldp@0...
Executing reset-failed of service lldp@1...
Executing reset-failed of service lldp@2...
Executing reset-failed of service lldp@3...
Executing reset-failed of service lldp@4...
Executing reset-failed of service lldp@5...
Executing reset-failed of service ntp-config...
Executing reset-failed of service pmon...
Executing reset-failed of service radv...
Executing reset-failed of service rsyslog-config...
Executing reset-failed of service snmp...
Executing reset-failed of service swss@1...
Executing reset-failed of service swss@0...
Executing reset-failed of service swss@2...
Executing reset-failed of service swss@3...
Executing reset-failed of service swss@4...
Executing reset-failed of service swss@5...
Executing reset-failed of service syncd@0...
Executing reset-failed of service syncd@1...
Executing reset-failed of service syncd@2...
Executing reset-failed of service syncd@3...
Executing reset-failed of service syncd@4...
Executing reset-failed of service syncd@5...
Executing reset-failed of service teamd@0...
Executing reset-failed of service teamd@1...
Executing reset-failed of service teamd@2...
Executing reset-failed of service teamd@3...
Executing reset-failed of service teamd@4...
Executing reset-failed of service teamd@5...
Executing restart of service hostname-config...
Executing restart of service interfaces-config...
Executing restart of service ntp-config...
Executing restart of service rsyslog-config...
Executing restart of service swss@0...
Executing restart of service swss@1...
Executing restart of service swss@2...
Executing restart of service swss@3...
Executing restart of service swss@4...
Executing restart of service swss@5...
Executing restart of service bgp@0...
Executing restart of service bgp@1...
Executing restart of service bgp@2...
Executing restart of service bgp@3...
Executing restart of service bgp@4...
Executing restart of service bgp@5...
Executing restart of service pmon...
Executing restart of service lldp...
Executing restart of service lldp@0...
Executing restart of service lldp@1...
Executing restart of service lldp@2...
Executing restart of service lldp@3...
Executing restart of service lldp@4...
Executing restart of service lldp@5...
Executing restart of service hostcfgd...

**real	1m55.971s**
user	0m3.652s
sys	0m1.232s

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

File: config/main.py [hunk 1]
```diff
@@ -132,14 +146,28 @@ def execute_systemctl(list_of_services, action):
...
-                    log_error("Failed to execute {} of service {}@{} with error {}".format(action, service, inst, e))
-                    raise
+            multi_inst_service_list.append(service)
+
+    # With Multi NPU, Start a thread per instance to do the "action" on multi instance services.
+    if sonic_device_util.is_multi_npu():
+        threads = []
+        # Use this event object to co-ordinate if any threads raised exception
+        e = threading.Event()
+        kwargs = {'multi_inst_list' : multi_inst_service_list, 'action' : action}
```
Comment (jleveque) [inline config/main.py:158]: Delete space before each `:`

File: config/main.py [hunk 2]
```diff
@@ -132,14 +146,28 @@ def execute_systemctl(list_of_services, action):
...
+        for inst in range(num_asic):
+            t = threading.Thread(target=execute_asic_instance, args=(inst,e), kwargs=kwargs)
```
Comment (jleveque) [inline config/main.py:160]: Ad space after comma in args tuple: `args=(inst, e)`

File: config/main.py (hunk 1 above)
Comment (judyjoseph) [reply config/main.py]: Updated the code.

File: config/main.py (hunk 2 above)
Comment (judyjoseph) [reply config/main.py]: Updated the code.

**Comment by judyjoseph** [comment]
retest vsimage please

**Comment by judyjoseph** [comment]
retest this please

**Comment by judyjoseph** [comment]
retest this please

File: config/main.py [hunk 3]
```diff
@@ -115,6 +116,16 @@ def get_command(self, ctx, cmd_name):
 # Helper functions
 #
 
+# Execute action per NPU instance for multi instance services.
+def execute_asic_instance(inst, event, service, action):
```
Comment (jleveque) [inline config/main.py:120]: Function name isn't descriptive enough. I believe it should include 'systemctl'.

Comment (judyjoseph) [reply config/main.py]: Updated the function name.

**Comment by jleveque** [comment]
Retest this please

**Comment by judyjoseph** [comment]
retest this please

**Comment by judyjoseph** [comment]
retest this please

**Comment by judyjoseph** [comment]
retest this please

**Comment by judyjoseph** [comment]
retest this please

**Comment by judyjoseph** [comment]
@jleveque - can you check this PR , this was blocked for a while due to test failures -- the tests are all passing now

**Pull request merged**